funsor/adjoint.py

```diff
@@ -28,10 +41,12 @@ def __init__(self):
         self._old_interpretation = None
 
     def __call__(self, cls, *args):
-        with interpretation(self._old_interpretation):
-            result = cls(*args)
-        if issubclass(cls, (Reduce, Contraction, Binary, Tensor)):  # TODO make generic
+        if cls in adjoint_ops:  # atomic op, don't trace internals
+            with interpretation(self._old_interpretation):
+                result = cls(*args)
             self.tape.append((result, cls, args))
+        else:
+            result = self._old_interpretation(cls, *args)
```

Comment:
FYI this change fixed the issues with `MarkovProduct` - the internal operations weren't being recorded